# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: third_party/rust/wgpu-core/src/init_tracker/buffer.rs
# Commit: ef5dc3e04e5f
# Full Hash: ef5dc3e04e5f271eea0636ab3a495e95cc912f1d
# Author: Dzmitry Malyshau <dmalyshau@mozilla.com>
# Date: 2021-09-04 09:40:24
# Regressor Bug: 1726626
# File Overlap Count: 3
# Description:
#   Bug 1726626 - Move gfx/wgpu into a 3rd party dependency r=jgilbert,bholley
#   
#   This update makes wgpu a vendored dependency instead of having it in gfx/wgpu.
#   
#   ## Notes
# ==============================================================================

diff --git a/third_party/rust/wgpu-core/src/init_tracker/buffer.rs b/third_party/rust/wgpu-core/src/init_tracker/buffer.rs
new file mode 100644
--- /dev/null
+++ b/third_party/rust/wgpu-core/src/init_tracker/buffer.rs
@@ -0,0 +1,33 @@
+use super::{InitTracker, MemoryInitKind};
+use crate::id::BufferId;
+use std::ops::Range;
+
+#[derive(Debug, Clone)]
+pub(crate) struct BufferInitTrackerAction {
+    pub id: BufferId,
+    pub range: Range<wgt::BufferAddress>,
+    pub kind: MemoryInitKind,
+}
+
+pub(crate) type BufferInitTracker = InitTracker<wgt::BufferAddress>;
+
+impl BufferInitTracker {
+    /// Checks if an action has/requires any effect on the initialization status and shrinks its range if possible.
+    pub(crate) fn check_action(
+        &self,
+        action: &BufferInitTrackerAction,
+    ) -> Option<BufferInitTrackerAction> {
+        self.create_action(action.id, action.range.clone(), action.kind)
+    }
+
+    /// Creates an action if it would have any effect on the initialization status and shrinks the range if possible.
+    pub(crate) fn create_action(
+        &self,
+        id: BufferId,
+        query_range: Range<wgt::BufferAddress>,
+        kind: MemoryInitKind,
+    ) -> Option<BufferInitTrackerAction> {
+        self.check(query_range)
+            .map(|range| BufferInitTrackerAction { id, range, kind })
+    }
+}